openqaoa/workflows/optimizer.py

```diff
@@ -430,42 +458,149 @@ def compile(self, problem: QUBO = None, verbose: bool = True):
-                steps=self.rqaoa_parameters.steps,
-                device=self.qaoa.device,
-                params_type=self.qaoa.circuit_properties.param_type,
-                init_type=self.qaoa.circuit_properties.init_type,
-                optimizer_dict=self.qaoa.classical_optimizer.asdict(),
-                backend_properties=self.qaoa.backend_properties.__dict__)
+
+        # reset all final variables
+        self.elimination_tracker = []
+        self.qaoa_steps = []
+        self.results = {}
+
+        # get variables
+        problem = self.problem  
+        n_cutoff = self.rqaoa_parameters.n_cutoff
+        n_qubits = problem.n
+        counter = self.rqaoa_parameters.counter 
+
+        # create a different max_terms function for each type 
+        if self.rqaoa_parameters.rqaoa_type.lower() == "adaptive":
+            f_max_terms = rqaoa.ada_max_terms  
         else:
-            raise f'rqaoa_type {self.rqaoa_parameters.rqaoa_type} is not supported. Please selet either "adaptive" or "custom'
+            f_max_terms = rqaoa.max_terms #? rqaoa.max_terms -> rqaoa.custom_max_terms ?
+
+        # If above cutoff, loop quantumly, else classically
+        while n_qubits > n_cutoff:
+
+            # Run QAOA
+            self._q.optimize()
+
+            # Obtain statistical results
+            exp_vals_z, corr_matrix = self._exp_val_hamiltonian_termwise()
+            # Retrieve highest expectation values according to adaptive method
+            max_terms_and_stats = f_max_terms(exp_vals_z, corr_matrix, self._n_step(n_qubits, n_cutoff, counter))
+            # Generate spin map
+            spin_map = rqaoa.spin_mapping(problem, max_terms_and_stats)
+            # Eliminate spins and redefine problem
+            new_problem, spin_map = rqaoa.redefine_problem(problem, spin_map)
+
+            # Extract final set of eliminations with correct dependencies and update tracker
+            eliminations = {(spin_map[spin][1],spin):spin_map[spin][0] for spin in sorted(spin_map.keys()) if spin != spin_map[spin][1]}
+            self.elimination_tracker.append(eliminations)
+
+            # Extract new number of qubits
+            n_qubits = new_problem.n
+
+            # Save qaoa object
+            self.qaoa_steps.append(copy.deepcopy(self._q))
+
+            # Compile qaoa with the new problem
+            self._q.compile(new_problem, verbose=False)
+            problem = new_problem
+
+            # Add one step to the counter
+            counter += 1
+
+        # Solve the new problem classically
+        cl_energy, cl_ground_states = ground_state_hamiltonian(problem.hamiltonian)
+
+        # Retrieve full solutions including eliminated spins and their energies
+        full_solutions = rqaoa.final_solution(
+            self.elimination_tracker, cl_ground_states, self.problem.hamiltonian)
+
+        # Compute description dictionary containing all the information                
```

Comment:
Answering a question you posed: I think it make sense to have these things formatted nicely and to create a rqaoa-result object (mimicking the structure of the QAOA one)

Reply:
I have created a RQAOAResults class to do so in my last commit, but for now is very different to the QAOA Results, I will change it.

But I am not sure where this class should be, for now is in optimizer.py, since the actual result.py file is in the classical optimizers methods.

Reply:
In this object, I have included a method: `get_qaoa_steps`, which returns a list of the qaoa objects that have been used in each step of the RQAOA

And also `get_elimination_rules`, which returns a list that keeps track of each elimination that has occurred.

Reply:
I have changed again the RQAOAResults class. I think now it makes more sense. The class inherits from `dict`, therefore `RQAOA.results` is a dictionary itself, but we can also create all the methods that we want. E.g. `RQAOAResults .get_hamiltonian_step(I)`, which returns the hamiltonian of the i-th intermediate QAOA step. 

```
r = RQAOA()
r.complie(problem)
r.optimize()
```
```
r.results
```
> {'solution': {'001011': -7.0, '110100': -7.0},
>  'classical_output': {'minimum_energy': -4.0,
>   'optimal_states': ['000', '111']},
>  'elimination_rules': [{(0, 2): -1.0}, {(1, 3): -1.0}, {(0, 3): -1.0}],
>  'schedule': [1, 1, 1],
>  'intermediate_steps': [{'QUBO': <openqaoa.problems.problem.QUBO at 0x282d0d93520>,
>    'QAOA': <openqaoa.workflows.optimizer.QAOA at 0x282d0d93130>},
>   {'QUBO': <openqaoa.problems.problem.QUBO at 0x282d160e070>,
>    'QAOA': <openqaoa.workflows.optimizer.QAOA at 0x282d15c6310>},
>   {'QUBO': <openqaoa.problems.problem.QUBO at 0x282d20a31c0>,
>    'QAOA': <openqaoa.workflows.optimizer.QAOA at 0x282d0c09c70>}],
>  'number_steps': 3}
```
r.results['solution']
```
>{'001011': -7.0, '110100': -7.0}
```
r.results.get_hamiltonian_step(2)
```
><openqaoa.qaoa_parameters.operators.Hamiltonian at 0x282d139bfa0>

Also, I have erased all the getters that were trivial. And now `r.results['intermediate_steps' ]` is a list of the intermediate QUBO and QAOA.

Let me know what you think.

Reply:
I like it!

My only comment is that I would want to have a JSON serialisable version of it already built in.

This current result can (perhaps) be saved with `pickle` but I am not 100% sure about it.

Either way, it would be good to have something like `r.results.dump()`, which:
- converts the object into a json
- converts the QUBO and the QAOA object into json too

Probably, the second step should be done elsewhere. That is, when you do `r.result.dump()`, then the `QUBO.dump()` and `QAOA.dump()` are called

Reply:
I was trying to do that by first converting all attributes to a dictionary by using the `convert2serialize` function but I get an error:
```
RecursionError: maximum recursion depth exceeded in comparison
```
It comes from when trying to convert the QAOA.backend: `convert2serialize(QAOA.backend)`

Should I use this function or create another one that is not recursive?

Reply:
Yes. And I don't think that one can be serialised easily (and probably it shouldn't).

Ok, let's not do too many things all at once. Let's first close this PR and then move on and make first the important parts of the QAOA object serialisable, and then let's deal with the RQAOA. What do you say?

Reply:
Okay, yes, that makes sense!

So, do I implement the r.result.dump() method but leaving the QAOA and QUBO objects not converted or I don't implement the method at all, for now?

Reply:
Leave it like this at the moment, and let's do a another PR right after this one to fix the formatting. It is an important feature to have!